File: ios/HybridTranslator.swift
import Foundation
import NitroModules
import MLKitTranslate

/// Nitro HybridObject implementation for on-device text translation using ML Kit.
/// Downloads language models on demand and caches the translator for a given language pair.
class HybridTranslator: HybridTranslatorSpec {

  // MARK: - HybridTranslatorSpec

  func translate(text: String, from: String, to: String) throws -> Promise<String> {
    guard let sourceLanguage = translateLanguage(from: from),
          let targetLanguage = translateLanguage(from: to) else {
      return Promise.rejected(withError: NSError(
        domain: "HybridTranslator", code: 1,
        userInfo: [NSLocalizedDescriptionKey: "Unsupported language pair: '\(from)' → '\(to)'"]))
    }

    let options = TranslatorOptions(sourceLanguage: sourceLanguage, targetLanguage: targetLanguage)
    let translator = Translator.translator(options: options)
    let sourceModel = TranslateRemoteModel.translateRemoteModel(language: sourceLanguage)
    let targetModel = TranslateRemoteModel.translateRemoteModel(language: targetLanguage)
    let promise = Promise<String>()

    // Download models if needed, then translate.
    // If translate returns code 13 (model files not found / corrupted), follow
    // ML Kit's own recovery advice: delete both models and re-download once.
    translator.downloadModelIfNeeded { error in
      if let error = error {
        promise.reject(withError: error)
        return
      }

      translator.translate(text) { translatedText, translateError in
        if let translateError = translateError {
          let nsError = translateError as NSError
          if nsError.domain == "com.google.mlkit" && nsError.code == 13 {
            // Model files missing/stale — delete and re-download once, then retry.
            let mm = ModelManager.modelManager()
            mm.deleteDownloadedModel(sourceModel) { _ in
              mm.deleteDownloadedModel(targetModel) { _ in
                translator.downloadModelIfNeeded { retryDownloadError in
                  if let retryDownloadError = retryDownloadError {
                    promise.reject(withError: retryDownloadError)
                    return
                  }
                  translator.translate(text) { retryText, retryError in
                    if let retryError = retryError {
                      promise.reject(withError: retryError)
                    } else {
                      promise.resolve(withResult: retryText ?? "")
                    }
                  }
                }
              }
            }
          } else {
            promise.reject(withError: translateError)
          }
        } else {
          promise.resolve(withResult: translatedText ?? "")
        }
      }
    }

    return promise
  }

  func removeLanguageModel(languageCode: String) throws -> Promise<Bool> {
    guard let language = translateLanguage(from: languageCode) else {
      return Promise.resolved(withResult: false)
    }

    let model = TranslateRemoteModel.translateRemoteModel(language: language)
    let modelManager = ModelManager.modelManager()

    guard modelManager.isModelDownloaded(model) else {
      return Promise.resolved(withResult: false)
    }

    let promise = Promise<Bool>()
    modelManager.deleteDownloadedModel(model) { error in
      if let error = error {
        promise.reject(withError: error)
      } else {
        promise.resolve(withResult: true)
      }
    }
    return promise
  }
}

File: android/src/main/java/com/margelo/nitro/visioncameraocrplus/HybridTranslator.kt
package com.margelo.nitro.visioncameraocrplus

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.common.model.DownloadConditions
import com.google.mlkit.common.model.RemoteModelManager
import com.google.mlkit.nl.translate.TranslateLanguage
import com.google.mlkit.nl.translate.TranslateRemoteModel
import com.google.mlkit.nl.translate.Translation
import com.google.mlkit.nl.translate.TranslatorOptions
import com.margelo.nitro.core.Promise

/**
 * Android implementation of the Translator HybridObject.
 * Uses ML Kit on-device translation with automatic model downloads.
 */
@DoNotStrip
@Keep
class HybridTranslator : HybridTranslatorSpec() {

  private val downloadConditions = DownloadConditions.Builder()
    .requireWifi()
    .build()

  override fun translate(text: String, from: String, to: String): Promise<String> {
    return Promise.async {
      val sourceTag = toTranslateLanguage(from)
        ?: throw RuntimeException("Unsupported source language: '$from'")
      val targetTag = toTranslateLanguage(to)
        ?: throw RuntimeException("Unsupported target language: '$to'")

      val options = TranslatorOptions.Builder()
        .setSourceLanguage(sourceTag)
        .setTargetLanguage(targetTag)
        .build()
      val translator = Translation.getClient(options)

      try {
        // Download model if needed (no-op when already present)
        Tasks.await(translator.downloadModelIfNeeded(downloadConditions))
        Tasks.await(translator.translate(text))
      } finally {
        translator.close()
      }
    }
  }

  override fun removeLanguageModel(languageCode: String): Promise<Boolean> {
    return Promise.async {
      val tag = toTranslateLanguage(languageCode) ?: return@async false
      val model = TranslateRemoteModel.Builder(tag).build()
      val modelManager = RemoteModelManager.getInstance()
      val isDownloaded = Tasks.await(modelManager.isModelDownloaded(model))
      if (!isDownloaded) return@async false
      Tasks.await(modelManager.deleteDownloadedModel(model))
      true
    }
  }

  // ---------------------------------------------------------------------------
  // Language code mapping  (BCP-47 → ML Kit TranslateLanguage tag)
  // ML Kit uses the same BCP-47 codes as the iOS MLKitTranslate constants.
  // ---------------------------------------------------------------------------
  private fun toTranslateLanguage(code: String): String? {
    return when (code.lowercase()) {
      "af" -> TranslateLanguage.AFRIKAANS
      "sq" -> TranslateLanguage.ALBANIAN
      "ar" -> TranslateLanguage.ARABIC
      "be" -> TranslateLanguage.BELARUSIAN
      "bn" -> TranslateLanguage.BENGALI
      "bg" -> TranslateLanguage.BULGARIAN
      "ca" -> TranslateLanguage.CATALAN
      "zh" -> TranslateLanguage.CHINESE
      "cs" -> TranslateLanguage.CZECH
      "da" -> TranslateLanguage.DANISH
      "nl" -> TranslateLanguage.DUTCH
      "en" -> TranslateLanguage.ENGLISH
      "eo" -> TranslateLanguage.ESPERANTO
      "et" -> TranslateLanguage.ESTONIAN
      "fi" -> TranslateLanguage.FINNISH
      "fr" -> TranslateLanguage.FRENCH
      "gl" -> TranslateLanguage.GALICIAN
      "ka" -> TranslateLanguage.GEORGIAN
      "de" -> TranslateLanguage.GERMAN
      "el" -> TranslateLanguage.GREEK
      "gu" -> TranslateLanguage.GUJARATI
      "ht" -> TranslateLanguage.HAITIAN_CREOLE
      "he" -> TranslateLanguage.HEBREW
      "hi" -> TranslateLanguage.HINDI
      "hu" -> TranslateLanguage.HUNGARIAN
      "is" -> TranslateLanguage.ICELANDIC
      "id" -> TranslateLanguage.INDONESIAN
      "ga" -> TranslateLanguage.IRISH
      "it" -> TranslateLanguage.ITALIAN
      "ja" -> TranslateLanguage.JAPANESE
      "kn" -> TranslateLanguage.KANNADA
      "ko" -> TranslateLanguage.KOREAN
      "lv" -> TranslateLanguage.LATVIAN
      "lt" -> TranslateLanguage.LITHUANIAN
      "mk" -> TranslateLanguage.MACEDONIAN
      "ms" -> TranslateLanguage.MALAY
      "mt" -> TranslateLanguage.MALTESE
      "mr" -> TranslateLanguage.MARATHI
      "no" -> TranslateLanguage.NORWEGIAN
      "fa" -> TranslateLanguage.PERSIAN
      "pl" -> TranslateLanguage.POLISH
      "pt" -> TranslateLanguage.PORTUGUESE
      "ro" -> TranslateLanguage.ROMANIAN
      "ru" -> TranslateLanguage.RUSSIAN
      "sk" -> TranslateLanguage.SLOVAK
      "sl" -> TranslateLanguage.SLOVENIAN
      "es" -> TranslateLanguage.SPANISH
      "sw" -> TranslateLanguage.SWAHILI
      "tl" -> TranslateLanguage.TAGALOG
      "ta" -> TranslateLanguage.TAMIL
      "te" -> TranslateLanguage.TELUGU
      "th" -> TranslateLanguage.THAI
      "tr" -> TranslateLanguage.TURKISH
      "uk" -> TranslateLanguage.UKRAINIAN
      "ur" -> TranslateLanguage.URDU
      "vi" -> TranslateLanguage.VIETNAMESE
      "cy" -> TranslateLanguage.WELSH
      else -> null
    }
  }
}
